### PR TITLE
[SPARK-12904][SQL] Strength reduction for integral and decimal literal comparisons

### PR DESCRIPTION
**Pull request by rxin**
This pull request implements strength reduction for comparing integral expressions and decimal literals, which is more common now because we switch to parsing fractional literals as decimal types (rather than doubles). I added the rules to the existing DecimalPrecision rule with some refactoring to simplify the control flow. I also moved DecimalPrecision rule into its own file due to the growing size.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  }
+
+  private def promotePrecision(e: Expression, dataType: DataType): Expression = {
+    PromotePrecision(Cast(e, dataType))
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    // fix decimal precision for expressions
+    case q => q.transformExpressions(
+      decimalAndDecimal.orElse(integralAndDecimalLiteral).orElse(nondecimalAndDecimal))
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: i broke the previous monolithic decimal precision rule into 2 parts, and then added integralAndDecimalLiteral

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+   * 8. decimal_literal <= int_col => ceil(decimal_literal) <= int_col
+   *
+   * Note that technically this is an "optimization" and should go into the optimizer. However,
+   * by the time the optimizer runs, these comparison expressions would be pretty hard to pattern
+   * match because there are multuple (at least 2) levels of casts involved.
+   *
+   * There are a lot more possible rules we can implement, but we don't do them
+   * because we are not sure how common they are.
+   */
+  private val integralAndDecimalLiteral: PartialFunction[Expression, Expression] = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: this is the only function in this file that is new. everything else was simply copied over from the existing DeicmalPrecision.

**Comment by rxin** [comment]
This should subsume https://github.com/apache/spark/pull/10845

cc @viirya - I refactored the code so it is a lot cleaner, and looked at overflow cases.

**Comment by viirya** [comment]
LGTM

**Comment by rxin** [comment]
Merging this in master.

**Pull request closed without merging**